digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/util/TrafficSignLinearGenerator.scala

```diff
@@ -1020,4 +1098,300 @@ class TrafficSignParkingProhibitionGenerator(roadLinkServiceImpl: RoadLinkServic
+      else if (trafficSign.expired) true
+      else compareValue(asset.value.get, createdValue.get)
+    }
+
+    toDelete.foreach { asset =>
+      linearAssetService.expireAsset(assetType, asset.id, username, true, false)
+      oracleLinearAssetDao.expireConnectedByLinearAsset(asset.id)
+    }
+
+    assetToUpdate(toUpdate, trafficSign, createdValue.get, userUpdate)
+  }
+
+  override def assetToUpdate(assets: Seq[PersistedLinearAsset], trafficSign: PersistedTrafficSign, createdValue: Value, username: String) : Unit = {
+    if (debbuger) println("assetToUpdate")
+    val groupedAssetsToUpdate = assets.map { asset =>
+      val newProperties = asset.value.get.asInstanceOf[DynamicValue].value.properties.find(_.publicId == "tyon_tunnus") ++ createdValue.asInstanceOf[DynamicValue].value.properties
+
+      (asset.id, newProperties)// asset.value.get.asInstanceOf[DynamicValue].value.properties.diff(createdValue.asInstanceOf[DynamicValue].value.properties))
+    }.groupBy(_._2)
+
+    groupedAssetsToUpdate.values.foreach { value =>
+      roadWorkService.updateWithoutTransaction(value.map(_._1), DynamicValue(DynamicAssetValue(value.flatMap(_._2))), username)
+      oracleLinearAssetDao.expireConnectedByPointAsset(trafficSign.id)
+    }
+  }
+
+  override def createLinearAsset(newSegment: TrafficSignToLinear, username: String) : Long = {
+    if (debbuger) println("createLinearAsset")
+    roadWorkService.createWithoutTransaction(assetType, newSegment.roadLink.linkId, newSegment.value,
+      newSegment.sideCode.value, Measures(newSegment.startMeasure, newSegment.endMeasure), username,
+      vvhClient.roadLinkData.createVVHTimeStamp(), Some(newSegment.roadLink) )
+  }
+
+  override def createSegmentPieces(actualRoadLink: RoadLink, allRoadLinks: Seq[RoadLink], sign: PersistedTrafficSign, signs: Seq[PersistedTrafficSign], pointOfInterest: (Option[Point], Option[Point], Option[Int]), result: Seq[TrafficSignToLinear]): Set[TrafficSignToLinear] = {
+    if (debbuger) println("createSegmentPieces")
+    createValue(Seq(sign)) match {
+      case Some(value) =>
+        val stopCondition = getStopCondition(actualRoadLink, sign, signs, pointOfInterest._3.get, result)
+        val generatedSegmentPieces = generateSegmentPiece(actualRoadLink, sign, value, stopCondition, pointOfInterest._3.get)
+
+        (if (stopCondition.isEmpty) {
+          val adjRoadLinks = getAdjacents(pointOfInterest, allRoadLinks.filterNot(_.linkId == actualRoadLink.linkId))
+          if (adjRoadLinks.nonEmpty) {
+            adjRoadLinks.flatMap { case (newRoadLink, (nextFirst, nextLast, nextDirection)) =>
+              createSegmentPieces(newRoadLink, allRoadLinks.filterNot(_.linkId == newRoadLink.linkId), sign, signs, (nextFirst, nextLast, nextDirection), generatedSegmentPieces +: result)
+            }
+          } else
+            generatedSegmentPieces +: result
+        } else
+          generatedSegmentPieces +: result).toSet
+      case _ => Set()
+    }
+  }
```

Comment:
This method's totally equal with other already existing at TrafficSignParkingProhibitionGenerator class.

Since "TrafficSignParkingProhibitionGenerator" and "TrafficSignRoadWorkGenerator" extend from "TrafficSignDynamicAssetGenerator" let's try to put the method on father class and using on the sons.

Reply:
The createSegmentPieces in TrafficSignLinearGenerator is different.
And TrafficSignDynamicAssetGenerator don't have it

Reply:
After talking, changes made as requested